src/codegen/solana_accounts/account_management.rs

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
+use crate::codegen::{Builtin, Expression};
+use crate::sema::ast::{ArrayLength, Function, Namespace, StructType, Type};
+use crate::sema::solana_accounts::{DATA_ACCOUNT, SYSTEM_ACCOUNT};
+use num_bigint::BigInt;
+use num_traits::Zero;
+use solang_parser::pt::Loc;
+use std::collections::{HashSet, VecDeque};
+
+/// This function walks over the CFG and automates the account management, so developers do not need
+/// to do so. For instance, when calling 'new construct{address: addr}()', we construct the correct
+/// AccountMeta array with all the accounts the constructor needs.
+pub(crate) fn manage_contract_accounts(contract_no: usize, ns: &mut Namespace) {
+    let contract_functions = ns.contracts[contract_no].functions.clone();
+    for function_no in &contract_functions {
+        let cfg_no = ns.contracts[contract_no]
+            .all_functions
+            .get(function_no)
+            .cloned()
+            .unwrap();
+        traverse_cfg(
+            &mut ns.contracts[contract_no].cfg[cfg_no],
+            &ns.functions,
+            *function_no,
+        );
+    }
+}
+
+/// This function walks over the CFG to process its instructions for the account management.
+fn traverse_cfg(cfg: &mut ControlFlowGraph, functions: &[Function], ast_no: usize) {
+    if cfg.blocks.is_empty() {
+        return;
+    }
+
+    let mut queue: VecDeque<usize> = VecDeque::new();
+    let mut visited: HashSet<usize> = HashSet::new();
+    queue.push_back(0);
+    visited.insert(0);
+
+    while let Some(cur_block) = queue.pop_front() {
+        for instr in cfg.blocks[cur_block].instr.iter_mut() {
+            process_instruction(instr, functions, ast_no);
+        }
+
+        for edge in cfg.blocks[cur_block].edges() {
+            if !visited.contains(&edge) {
+                queue.push_back(edge);
+                visited.insert(edge);
+            }
+        }
+    }
+}
+
+/// This function processes the instruction, creating the AccountMeta array when possible.
+/// Presently, we only check the Instr::Constructor, but more will come later.
+fn process_instruction(instr: &mut Instr, functions: &[Function], ast_no: usize) {
+    if let Instr::Constructor {
+        accounts,
+        address,
+        constructor_no,
+        ..
+    } = instr
+    {
+        if accounts.is_some() || constructor_no.is_none() {
+            return;
+        }
+
+        let mut account_metas: Vec<Expression> = Vec::new();
+        let constructor_func = &functions[constructor_no.unwrap()];
+        for (name, account) in constructor_func.solana_accounts.borrow().iter() {
+            if name == DATA_ACCOUNT {
```

Comment:
These accounts might as well be an enum, and then you could implement this logic here (as well as to and from string implementations) on that enum